sdk/trace/provider.go

```diff
@@ -0,0 +1,169 @@
+// Copyright 2019, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package trace
+
+import (
+	"sync"
+	"sync/atomic"
+
+	"go.opentelemetry.io/sdk/export"
+
+	apitrace "go.opentelemetry.io/api/trace"
+)
+
+const (
+	defaultTracerName = "go.opentelemetry.io/sdk/tracer"
+)
+
+type ProviderOptions struct {
+	syncers  []export.SpanSyncer
+	batchers []export.SpanBatcher
+	config   Config
+}
+
+type ProviderOption func(*ProviderOptions)
+
+type Provider struct {
+	mu             sync.Mutex
+	namedTracer    map[string]*tracer
+	spanProcessors atomic.Value
+	config         atomic.Value // access atomically
+}
+
+var _ apitrace.Provider = &Provider{}
+
+func NewProvider(opts ...ProviderOption) (*Provider, error) {
+
+	o := &ProviderOptions{}
+
+	for _, opt := range opts {
+		opt(o)
+	}
+
+	tp := &Provider{
+		namedTracer: make(map[string]*tracer),
+	}
+	tp.config.Store(&Config{
+		DefaultSampler:       ProbabilitySampler(defaultSamplingProbability),
+		IDGenerator:          defIDGenerator(),
+		MaxAttributesPerSpan: DefaultMaxAttributesPerSpan,
+		MaxEventsPerSpan:     DefaultMaxEventsPerSpan,
+		MaxLinksPerSpan:      DefaultMaxLinksPerSpan,
+	})
+
+	for _, syncer := range o.syncers {
+		ssp := NewSimpleSpanProcessor(syncer)
+		tp.RegisterSpanProcessor(ssp)
+	}
+
+	for _, batcher := range o.batchers {
+		bsp, err := NewBatchSpanProcessor(batcher)
+		if err != nil {
+			return nil, err
+		}
+		tp.RegisterSpanProcessor(bsp)
+	}
+
+	tp.ApplyConfig(o.config)
+
+	return tp, nil
+}
+
+func (p *Provider) GetTracer(name string) apitrace.Tracer {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	if name == "" {
+		name = defaultTracerName
+	}
+	t, ok := p.namedTracer[name]
+	if !ok {
+		t = &tracer{name: name, provider: p}
+		p.namedTracer[name] = t
+	}
+	return t
+}
+
+func (p *Provider) RegisterSpanProcessor(s SpanProcessor) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	new := make(spanProcessorMap)
+	if old, ok := p.spanProcessors.Load().(spanProcessorMap); ok {
+		for k, v := range old {
+			new[k] = v
+		}
+	}
+	new[s] = &sync.Once{}
+	p.spanProcessors.Store(new)
+}
+
+// UnregisterSpanProcessor removes from the list of SpanProcessors the SpanProcessor that was
+// registered with the given name.
```

Comment:
This comment is quite stuttery.
Suggestion:
`UnregisterSpanProcessor removes the given SpanProcessor from the list of SpanProcessors`